### ISSUE TITLE
Questions to the performance related to reuse the context in Esper 8 

### ISSUE THREAD

**Issue by RunzhongHuang** [espertechinc/esper#259]
Earlier in Esper 7, we can use following to declare a context and a statement separately.

```
String context = "create context contextName ....";
String statement = "context contextName select ....";
epAdministrator.createEPL(context , "contextName");
epAdministrator.createEPL(statement , "statementName");
```

Now in Esper 8. So far I followed some best practices as below.

```
String context = "create context contextName ....;\n";
String statement_1 = "@name('test1') context contextName select ....;\n";
EPCompiled compiled_full_statement_1 = EPCompilerProvider.getCompiler().compile(context + statement_1, new runtime.getDeploymentService().deploy(compiled_full_statement_1)

String statement_2 = "@name('test2') context contextName insert ....;\n";
EPCompiled compiled_full_statement_2 = EPCompilerProvider.getCompiler().compile(context + statement_2, new runtime.getDeploymentService().deploy(compiled_full_statement_2)
```

As you can see in Esper 8, whenever I want to compile & deploy a statement by reusing the existing context. I need to include "create context ... " statement again during compilation. So I am wondering:

- Will this cause performance issues? And how is this scenario handled internally?
- Is there a better implementation than the one I showed above in Esper 8?

I read the esper-reference document and didn't find a solid answer, thus I am looking for help from the community! Thank you!

**Comment by raguessner** [comment]
Yes there is a better implementation. Defining the same context multiple times may actually impact performance. Esper 8 is a compiler and a runtime, while Esper 7 and older was just a runtime with an interpreter. The Esper 8 compiler has a path which tells it what existing contexts, named windows, event types etc. there are. One of the other grapes some users had with old versions of Esper was that there was no real multitenancy as everything was visible to everything. Esper 8 fixed that with access modifiers. 

For reading up, there are some posts for migrating applications from older versions. The link is (https://www.espertech.com/2018/10/23/esper-8-migrating-step-by-step-for-existing-applications/) and documentation link is [compiler arguments](http://esper.espertech.com/release-8.8.0/reference-esper/html_single/index.html#apicompiler-compilerarguments-path)
